### PR TITLE
CIRC-937: Allow override due date when proposed due date is same or before current.

### PR DESCRIPTION
**Pull request by bohdan-suprun**
https://issues.folio.org/browse/CIRC-937 - BE: Override failed renewal when renewal will not change the due date.

## Approach
* Added an override case to allow override renew if proposed due date is the same or earlier to the current due date.
* Use override due date for all the existing override cases when proposed due date is the same or before the current.
* Require override due date for all the cases when proposed due date is the same or before the current.
* Do not allow declared lost case to break the validation.

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/test/java/org/folio/circulation/resources/RenewalValidatorTest.java [hunk 1]
```diff
@@ -18,35 +18,34 @@
...
 
     val validationResult = errorWhenEarlierOrSameDueDate(loan, proposedDueDate);
 
-    assertTrue(validationResult.succeeded());
+    assertTrue(validationResult.failed());
+    assertThat(validationResult.cause(), instanceOf(ValidationErrorFailure.class));
+
+    val validationFailure = (ValidationErrorFailure) validationResult.cause();
+    assertThat(validationFailure.hasErrorWithReason("renewal would not change the due date"),
+      is(true));
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/resources/RenewalValidatorTest.java:33]: Could this be extracted into a matcher to reduce the amount of lower level checks in this test?

File: src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+import org.folio.circulation.support.ValidationErrorFailure;
+import org.folio.circulation.support.results.Result;
+import org.joda.time.DateTime;
+import org.junit.Test;
+
+import api.support.builders.LoanPolicyBuilder;
+import io.vertx.core.json.JsonObject;
+
+public class OverrideRenewalStrategyTest {
+
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java:34]: Personally, I don't think this blank line improves the readability of the code

File: src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+      .changeDueDate(now(UTC).plusWeeks(1).plusSeconds(1))
+      .withLoanPolicy(loanPolicy);
+
+    final Result<Loan> renewedLoan = renew(loan, overrideDueDate);
+
+    assertDueDate(overrideDueDate, renewedLoan);
+  }
+
+  @Test
+  public void shouldUseProposedDueDateWhenReachedNumberOfRenewalsAndNewDueDateAfterCurrent() {
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java:125]: What is the difference between the proposed due date and the override due date, is the proposed the one determined automatically by the system?

File: src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+  private void assertValidationError(Result<Loan> renewResult, String expectedMessage) {
+    assertTrue(renewResult.failed());
+    assertTrue(renewResult.cause() instanceof ValidationErrorFailure);
+
+    final ValidationErrorFailure failure = (ValidationErrorFailure) renewResult.cause();
+    if (!failure.hasErrorWithReason(expectedMessage)) {
+      System.out.println("Expected: " + expectedMessage);
+      System.out.println("Actual: " + failure.getErrors().iterator().next().getMessage());
+    }
+    assertTrue(failure.toString().contains(expectedMessage));
```
Comment (marcjohnson-kint) [inline src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java:354]: This seems very similar to code in another class could these be combined into a matcher (or set of matchers)?

**Review by marcjohnson-kint — APPROVED** [review]

File: src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java (hunk 3 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java:125]: Yes, the proposed due date is date and time calculated by policy and override due date - due date specified by user in override renew request.

Comment (marcjohnson-kint) [reply src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java:125]: I think of proposed as the due date proposed by the client, maybe a term like `calculated` could make this distinction stronger, or alternatively a comment to explain the distinction?

File: src/test/java/org/folio/circulation/resources/RenewalValidatorTest.java (hunk 1 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/circulation/resources/RenewalValidatorTest.java]: Done, extracted to `hasValidationError` matcher.

File: src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java (hunk 3 above)
Comment (bohdan-suprun) [reply src/test/java/org/folio/circulation/resources/renewal/OverrideRenewalStrategyTest.java]: Make sense, I use `calculated` term now.

**Pull request merged**